File: extensions/shared-youtube/library/src/main/java/app/morphe/extension/shared/patches/FixRecycledBitmapPatch.java
/*
 * Copyright 2026 Morphe.
 * https://github.com/MorpheApp/morphe-patches
 *
 * See the included NOTICE file for GPLv3 Section 7 terms that apply to this code.
 */

package app.morphe.extension.shared.patches;

import android.graphics.Bitmap;
import android.graphics.Bitmap.Config;
import android.media.MediaMetadata.Builder;

import app.morphe.extension.shared.Logger;

@SuppressWarnings("unused")
public class FixRecycledBitmapPatch {

    /**
     * Injection point.
     * Fix: <a href="https://github.com/MorpheApp/morphe-patches/issues/686">Cannot obtain size for recycled Bitmap: ARGB_8888</a>.
     */
    public static Builder putBitmap(Builder builder, String key, Bitmap bitmap) {
        if (bitmap != null) {
            if (bitmap.isRecycled()) {
                Logger.printDebug(() -> "Bitmap is recycled, creating a dummy bitmap");
                bitmap = Bitmap.createBitmap(0, 0, Config.ARGB_8888);
            }

            try {
                Config config = bitmap.getConfig();
                if (config == null) config = Config.ARGB_8888;
                builder = builder.putBitmap(key, bitmap.copy(config, false));
                Logger.printDebug(() -> "Bitmap copied");
            } catch (Exception ex) {
                Logger.printException(() -> "Failed to copy bitmap", ex);
            }
        }

        return builder;
    }
}

File: patches/src/main/kotlin/app/morphe/patches/shared/misc/fix/bitmap/FixRecycledBitmapPatch.kt
/*
 * Copyright 2026 Morphe.
 * https://github.com/MorpheApp/morphe-patches
 *
 * See the included NOTICE file for GPLv3 §7(b) and §7(c) terms that apply to this code.
 */

package app.morphe.patches.shared.misc.fix.bitmap

import app.morphe.patcher.Fingerprint
import app.morphe.patcher.extensions.InstructionExtensions.replaceInstruction
import app.morphe.patcher.methodCall
import app.morphe.patcher.patch.bytecodePatch
import app.morphe.util.fiveRegisters
import app.morphe.util.matchAllMethodIndicesForEach

private const val EXTENSION_CLASS =
    "Lapp/morphe/extension/shared/patches/FixRecycledBitmapPatch;"

val fixRecycledBitmapPatch = bytecodePatch(
    description = "Fixes recycled bitmap crashes by routing putBitmap through the extension class."
) {

    execute {
        Fingerprint(
            filters = listOf(
                methodCall(
                    definingClass = $$"Landroid/media/MediaMetadata$Builder;",
                    name = "putBitmap",
                    parameters = listOf("Ljava/lang/String;", "Landroid/graphics/Bitmap;")
                )
            ),
            custom = { _, classDef ->
                !classDef.type.startsWith("Lapp/morphe/extension")
            }
        ).matchAllMethodIndicesForEach(requireMatches = false) { index ->
            val registers = fiveRegisters(index)

            replaceInstruction(
                index,
                $"invoke-static { $registers }, $EXTENSION_CLASS->putBitmap(" +
                        "Landroid/media/MediaMetadata\$Builder;Ljava/lang/String;Landroid/graphics/Bitmap;)" +
                        "Landroid/media/MediaMetadata\$Builder;"
            )
        }
    }
}